modules/apps/transfer/ibc_module_test.go

```diff
@@ -403,3 +403,76 @@ func (suite *TransferTestSuite) TestOnChanUpgradeTry() {
 		})
 	}
 }
+
+func (suite *TransferTestSuite) TestOnChanUpgradeAck() {
+	var path *ibctesting.Path
+
+	testCases := []struct {
+		name     string
+		malleate func()
+		expError error
+	}{
+		{
+			"success",
+			func() {}, // successful happy path for a standalone transfer app is swapping out the underlying connection
+			nil,
+		},
+		{
+			"invalid upgrade version",
+			func() {
+				// NOTE: counterpartyUpgrade is queried by Endpoint.ChanUpgradeAck() so retrieve and mutate fields here appropriately to force failure
+				counterpartyUpgrade := path.EndpointB.GetChannelUpgrade()
+				counterpartyUpgrade.Fields.Version = "invalid-version"
+				path.EndpointB.SetChannelUpgrade(counterpartyUpgrade)
+
+				suite.coordinator.CommitBlock(suite.chainB)
+			},
+			channeltypes.NewUpgradeError(1, types.ErrInvalidVersion),
+		},
+	}
+
+	for _, tc := range testCases {
+		tc := tc
+		suite.Run(tc.name, func() {
+			suite.SetupTest()
+
+			path = NewTransferPath(suite.chainA, suite.chainB)
+			suite.coordinator.Setup(path)
+
+			// configure the channel upgrade to modify the underlying connection
+			upgradePath := ibctesting.NewPath(suite.chainA, suite.chainB)
+			suite.coordinator.SetupConnections(upgradePath)
+
+			path.EndpointA.ChannelConfig.ProposedUpgrade.Fields.ConnectionHops = []string{upgradePath.EndpointA.ConnectionID}
+			path.EndpointB.ChannelConfig.ProposedUpgrade.Fields.ConnectionHops = []string{upgradePath.EndpointB.ConnectionID}
+
+			err := path.EndpointA.ChanUpgradeInit()
+			suite.Require().NoError(err)
+
+			err = path.EndpointB.ChanUpgradeTry()
+			suite.Require().NoError(err)
+
+			tc.malleate()
+
+			err = path.EndpointA.ChanUpgradeAck()
```

Comment:
interesting that you call the endpoint to obtain the error. Is this going to be problematic if we make changes to other module in the middleware stack? I wonder if we should just call the module's `OnChanUpgradeAck` directly, it would simplify this test code as well as we could remove assertions on what core IBC should be doing

Reply:
I've updated to invoke the transfer callback directly on the method under test. `Endpoint` methods are still used for setup (INIT, TRY). 

I can see arguments for having both ways, direct call vs full stack with `Endpoint` call. 

We can keep using the full app stack calls in https://github.com/cosmos/ibc-go/pull/4129 but sure, it might be more maintainable to test callbacks directly here.

Reply:
Refactored `OnChanUpgradeTry` tests here - https://github.com/cosmos/ibc-go/pull/4221

Testing for `OnChanUpgradeInit` can remain the same I think!